aws/callback.ts

```diff
@@ -0,0 +1,113 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as aws from "@pulumi/aws";
+import * as cloud from "@pulumi/cloud";
+import * as pulumi from "@pulumi/pulumi";
+import { functionIncludePackages, functionIncludePaths, functionMemorySize } from "./config";
+import { getComputeIAMRolePolicies, getOrCreateNetwork, runLambdaInVPC } from "./shared";
+
+/**
+ * AWS-specific data to create an AWS lambda out of a callback function.  Can be passed to any
+ * functions in pulumi/cloud-aws that can take a cloud.Callback<T> argument.  To create the same
+ * default AwsCallbackData Pulumi creates when given a simple JavaScript function, use
+ * [createCallbackData].
+ */
+export interface AwsCallbackData<T extends Function> extends cloud.CallbackData<T>, aws.lambda.CallbackFunctionArgs<any, any> {
+    /**
+     * Not used.  Provide [function] instead.
+     */
+    callback?: never;
+    /**
+     * Not used.  Provide [function] instead.
+     */
+    callbackFactory?: never;
+}
```

Comment:
This is an interesting way things ended up landing.  We want to derive from CallbackFunctionArgs so that the user can provide any and all properties that may flow into an aws.lamdba.CallbackFunction.  However, we don't want them specifically providing 'callback' or 'callbackFactory'.

These two ar ethe function the pulumi/cloud layer creates *on behalf of* the user.  i.e. the user provides some sensible pulumi/cloud-oriented function instance, and we wrap that with the right shapes here and provide the callback/callbackFactory for them.

Specifically, the shape of these helpers is *completely* unrelated to the shape of the cloud-callback function tehy provide.  For example, for Bucket, the shape of the function the usre providers takes in cloud.Bucket related types.  whereas these functions are the ones that form the entrypoint of hte AWS lambda (i.e. the context,  and callback objects).

Note: i'm nto a fan of having to do thse 'nevers' here.  A cleaner way would be to break the core API up at the aws/azure layer and have someting like:

```ts
// in pulumi/aws

interface BaseFunctionArgs = utils.Overwrite<FunctionArgs, {
    ...
}>

interface CallbackFunctionArgs<E, R> extends BaseFunctionArgs {
    callback?:  Callback<E, R>;
    callbackFactory?: CallbackFactory<E, R>;
}

// in pulumi/cloud-aws:

interface AwsCallbackData<T extends Function> extends cloud.CallbackData<T>, aws.lambda.BaseFunctionArgs {
}
```

Thoughts?